kani-compiler/src/codegen_aeneas_llbc/mir_to_ullbc/mod.rs

```diff
@@ -0,0 +1,799 @@
+};
+use charon_lib::{error_assert, error_or_panic};
+use rustc_errors::MultiSpan;
+use rustc_middle::ty::TyCtxt;
+use rustc_smir::rustc_internal;
+use rustc_span::def_id::DefId as InternalDefId;
+use stable_mir::abi::PassMode;
+use stable_mir::mir::mono::Instance;
+use stable_mir::mir::{
+    BasicBlock, BinOp, Body, BorrowKind, CastKind, ConstOperand, Mutability, Operand, Place,
+    ProjectionElem, Rvalue, Statement, StatementKind, SwitchTargets, Terminator, TerminatorKind,
+};
+use stable_mir::ty::{
+    Allocation, ConstantKind, IndexedVal, IntTy, MirConst, Region, RegionKind, RigidTy, Span, Ty,
+    TyKind, UintTy,
+};
+
+use stable_mir::{CrateDef, DefId};
+use tracing::{debug, trace};
+
+/// A context for translating a single MIR function to ULLBC.
+/// The results of the translation are stored in the `translated` field.
+pub struct Context<'a, 'tcx> {
+    tcx: TyCtxt<'tcx>,
+    instance: Instance,
+    translated: &'a mut TranslatedCrate,
+    errors: &'a mut ErrorCtx<'tcx>,
+}
+
+impl<'a, 'tcx> Context<'a, 'tcx> {
+    /// Create a new context for translating the function `instance`, populating
+    /// the results of the translation in `translated`
+    pub fn new(
+        tcx: TyCtxt<'tcx>,
+        instance: Instance,
+        translated: &'a mut TranslatedCrate,
+        errors: &'a mut ErrorCtx<'tcx>,
+    ) -> Self {
+        Self { tcx, instance, translated, errors }
+    }
+
+    fn tcx(&self) -> TyCtxt<'tcx> {
+        self.tcx
+    }
+
+    fn span_err<S: Into<MultiSpan>>(&mut self, span: S, msg: &str) {
+        self.errors.span_err(span, msg);
+    }
+
+    fn continue_on_failure(&self) -> bool {
+        self.errors.continue_on_failure()
+    }
+
+    /// Perform the translation
+    pub fn translate(&mut self) -> Result<(), ()> {
+        // Charon's `id_map` is in terms of internal `DefId`
+        let def_id = rustc_internal::internal(self.tcx(), self.instance.def.def_id());
+
+        // TODO: might want to populate `errors.dep_sources` to help with
+        // debugging
+
+        let fid = self.register_fun_decl_id(def_id);
+
+        let item_meta = match self.translate_item_meta_from_rid(self.instance) {
+            Ok(item_meta) => item_meta,
+            Err(_) => {
+                return Err(());
+            }
+        };
+
+        let signature = self.translate_function_signature();
+        let body = match self.translate_function_body() {
+            Ok(body) => body,
+            Err(_) => {
+                return Err(());
+            }
+        };
+
+        let fun_decl = FunDecl {
+            def_id: fid,
+            rust_id: def_id,
+            item_meta,
+            signature,
+            kind: ItemKind::Regular,
+            body: Ok(body),
+        };
+
+        self.translated.fun_decls.set_slot(fid, fun_decl);
+
+        Ok(())
+    }
+
+    /// Get or create a `FunDeclId` for the given function
+    fn register_fun_decl_id(&mut self, def_id: InternalDefId) -> FunDeclId {
```

Comment:
Why do you need internal id here?

Reply:
Charon's data structures are in terms of the internal `DefId`: https://github.com/AeneasVerif/charon/blob/4caeeb35331b7e6ee59565a3eb698a200c6e90e9/charon/src/ast/krate.rs#L110

This was removed recently (in https://github.com/AeneasVerif/charon/commit/8a80793e440b256611baf747483a1b24613a043e), so once we upgrade to that commit, we should no longer need it.